contrib/mesos/pkg/executor/service/service.go

```diff
@@ -178,7 +182,93 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
 		log.Warningf("Unknown Docker exec handler %q; defaulting to native", s.DockerExecHandlerName)
 		dockerExecHandler = &dockertools.NativeExecHandler{}
 	}
+	dockerClient := dockertools.ConnectToDockerOrDie(s.DockerEndpoint)
+
+	//TODO(jdef) either configure Watch here with something useful, or else
+	// get rid of it from executor.Config
+	kubeletFinished := make(chan struct{})
+	execUpdates := make(chan kubelet.PodUpdate, 1)
+	exec := executor.New(executor.Config{
+		Updates:         execUpdates,
+		APIClient:       apiclient,
+		Docker:          dockerClient,
+		SuicideTimeout:  s.SuicideTimeout,
+		KubeletFinished: kubeletFinished,
+		ShutdownAlert: func() {
+			if shutdownCloser != nil {
+				if e := shutdownCloser.Close(); e != nil {
+					log.Warningf("failed to signal shutdown to external watcher: %v", e)
+				}
+			}
+		},
+		ExitFunc: os.Exit,
+		PodStatusFunc: func(pod *api.Pod) (*api.PodStatus, error) {
```

Comment:
this func is what provides information to the apiserver.

On Tue, Sep 8, 2015 at 10:15 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r38930044>
> :
>
> > +	execUpdates := make(chan kubelet.PodUpdate, 1)
> > +	exec := executor.New(executor.Config{
> > +		Updates:         execUpdates,
> > +		APIClient:       apiclient,
> > +		Docker:          dockerClient,
> > +		SuicideTimeout:  s.SuicideTimeout,
> > +		KubeletFinished: kubeletFinished,
> > +		ShutdownAlert: func() {
> > +			if shutdownCloser != nil {
> > +				if e := shutdownCloser.Close(); e != nil {
> > +					log.Warningf("failed to signal shutdown to external watcher: %v", e)
> > +				}
> > +			}
> > +		},
> > +		ExitFunc: os.Exit,
> > +		PodStatusFunc: func(pod *api.Pod) (*api.PodStatus, error) {
>
> It is dirty. Explicitly dirty ;-) I want to get rid of this whole callback
> mechanism. Why can't we just use the apiserver to get this information?
>
> My mid-term goal is to run the executor and the kubelet without any need
> to synchronize their start-up sequence.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r38930044>.
>